src/mantine-core/src/Stepper/Stepper.tsx

```diff
@@ -117,10 +117,20 @@ export const Stepper: StepperComponent = forwardRef<HTMLDivElement, StepperProps
   const items = _children.reduce<React.ReactElement[]>((acc, item, index) => {
     const state =
       active === index ? 'stepProgress' : active > index ? 'stepCompleted' : 'stepInactive';
-    const shouldAllowSelect = state === 'stepCompleted' || allowNextStepsSelect;
-    typeof item.props.allowStepSelect === 'boolean'
-      ? item.props.allowStepSelect
-      : typeof onStepClick === 'function';
+
+    const shouldAllowSelect = () => {
+      if (typeof onStepClick !== 'function') {
+        return false;
+      }
+
+      if (typeof item.props.allowStepSelect === 'boolean') {
```

Comment:
Giving priority to the Step's allowStepSelect prop made sense for more fine grain control over the step selection.

Reply:
Indeed, and here was actually the typo I added in the previous code... I forgot a ` &&`  operator, with the "typeof item.props..." part staying alone on its own line for nothing.

Reply:
In terms of readability I agree with your change; in terms of performance since this is in a loop, the added function call might not be the best idea. Then again, I don't imagine a stepper component getting tens or hundreds of steps :)

Reply:
The following will achieve the same, without the function call:

```ts
    const defaultAllowSelect = state === 'stepCompleted' || allowNextStepsSelect;
    const shouldAllowSelect =
      defaultAllowSelect &&
      (typeof item.props.allowStepSelect === 'boolean'
        ? item.props.allowStepSelect
        : typeof onStepClick === 'function');
``` 

But it's arguably more difficult to read.

Reply:
The performance change is so negligible (even with a thousand steps) that it may as well be part of the margin of error and at that point I'd rather favor readability

Reply:
Yeah I agree. This was definitely a nitpick.